repository-azure/src/main/java/io/aiven/elasticsearch/repositories/azure/AzureClient.java

```diff
@@ -0,0 +1,85 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.elasticsearch.repositories.azure;
+
+import java.util.Objects;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import com.azure.core.http.okhttp.OkHttpAsyncHttpClientBuilder;
+import com.azure.core.http.policy.UserAgentPolicy;
+import com.azure.storage.blob.BlobServiceClient;
+import com.azure.storage.blob.BlobServiceClientBuilder;
+import com.azure.storage.common.policy.RequestRetryOptions;
+import com.azure.storage.common.policy.RetryPolicyType;
+import okhttp3.Dispatcher;
+
+class AzureClient implements AutoCloseable {
+
+    static final String HTTP_USER_AGENT = "aiven-azure-repository";
+
+    private final BlobServiceClient blobServiceClient;
+
+    private final ExecutorService httpPoolExecutorService;
+
+    private AzureClient(final BlobServiceClient blobServiceClient, final ExecutorService httpPoolExecutorService) {
+        this.blobServiceClient = blobServiceClient;
+        this.httpPoolExecutorService = httpPoolExecutorService;
+    }
+
+    public BlobServiceClient blobServiceClient() {
+        return blobServiceClient;
+    }
+
+    public static AzureClient create(final AzureStorageSettings azureStorageSettings) {
+        final var httpThreadPoolSettings = azureStorageSettings.httpThreadPoolSettings();
+        final var httpClientExecutor = new ThreadPoolExecutor(
+                httpThreadPoolSettings.minThreads(), httpThreadPoolSettings.maxThreads(),
+                httpThreadPoolSettings.keepAlive(), TimeUnit.SECONDS,
+                new SynchronousQueue<>()
+        );
+        final var blobServiceClient = new BlobServiceClientBuilder()
+                .connectionString(azureStorageSettings.azureConnectionString())
+                .retryOptions(
+                        new RequestRetryOptions(
+                                RetryPolicyType.EXPONENTIAL, azureStorageSettings.maxRetries(),
+                                null, null, null, null))
+                .addPolicy(new UserAgentPolicy(HTTP_USER_AGENT))
+                .httpClient(
+                        new OkHttpAsyncHttpClientBuilder()
+                                .dispatcher(new Dispatcher(httpClientExecutor))
+                                .build())
+                .buildClient();
+        return new AzureClient(blobServiceClient, httpClientExecutor);
+    }
+
+    @Override
+    public void close() {
+        if (Objects.nonNull(httpPoolExecutorService)) {
```

Comment:
Can it ever be null?

Reply:
Yes, in case unfinished configuration ...